### PR TITLE
1012 bug in getting testparameters in specific test derived from generictest

### LINKED ISSUE #1012: Bug in getting TestParameters in specific test derived from GenericTest
### Bug description

When calling get_default() from AntigenTest or PCRTest, you get the TestParameters number from get_default() of GenericTest. 

### Version

Any

### To reproduce

1. compile with cmake --build .
2. call get_default() from AntigenTest or PCRTest
3. get sensitivity or specificity from GenericTest

### Add any relevant information, e.g. used compiler, screenshots.

_No response_

### Checklist

- [X] Attached labels, especially loc:: or model:: labels.
- [X] Linked to project

### PR DESCRIPTION
**Pull request by khoanguyen-dev**
# Changes and Information

Please **briefly list the changes** (main added features, changed items, or corrected bugs) made:

- Change TestType to enum class
- Create TestData struct to get TestParameter for each type
- Add a test to test the get_default()
- Delete the GenericTest and other type
- Modify the previous tests to accommodate these changes

If need be, add additional information and what the reviewer should look out for in particular:

- The changes are mostly made in parameters.h
- The TestData is added as one of the World parameters and TestParameters are extracted when they are needed. 

## Merge Request - Guideline Checklist

Please check our [git workflow](https://github.com/SciCompMod/memilio/wiki/git-workflow). Use the **draft** feature if the Pull Request is not yet ready to review.

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/SciCompMod/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful (with and without OpenMP)
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added
- [x] (For ABM development) Checked [benchmark results](https://github.com/SciCompMod/memilio/wiki/Agent-Based-Model-Development) and ran and posted a local test above from before and after development to ensure performance is monitored.

### Checks by code reviewer(s)

- [x] Corresponding issue(s) is/are linked and addressed
- [x] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [x] Appropriate **unit tests** have been added, CI passes, code coverage and performance is acceptable (did not decrease)
- [x] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] On merge, add 2-5 lines with the changes (main added features, changed items, or corrected bugs) to the merge-commit-message. This can be taken from the **briefly-list-the-changes** above (best case) or the separate commit messages (worst case).

Closes #1012

### CODE REVIEW

File: cpp/models/abm/parameters.h [hunk 1]
```diff
@@ -316,45 +317,19 @@ struct TestParameters {
...
-    }
-};
-
-/**
- * @brief Reliability of a PCRTest.
- */
-struct PCRTest : public GenericTest {
-    using Type = TestParameters;
-    static Type get_default()
+struct TestData {
```
Comment (reneSchm) [inline cpp/models/abm/parameters.h:320]: Please add a brief description to the parameter.

File: cpp/tests/test_abm_testing_strategy.cpp [hunk 2]
```diff
@@ -150,3 +150,25 @@ TEST(TestTestingScheme, initAndRunTestingStrategy)
...
+    ASSERT_NEAR(test_parameters.sensitivity, 0.9, 1e-7);
+    ASSERT_NEAR(test_parameters.specificity, 0.99, 1e-7);
+
+    // Test whether the TestParameters of TestData can be modified.
+    world.parameters.get<mio::abm::TestData>()[mio::abm::TestType::PCR].sensitivity = 0.8;
+    world.parameters.get<mio::abm::TestData>()[mio::abm::TestType::PCR].specificity = 0.88;
+    test_parameters = world.parameters.get<mio::abm::TestData>()[mio::abm::TestType::PCR];
+    ASSERT_NEAR(test_parameters.sensitivity, 0.8, 1e-7);
+    ASSERT_NEAR(test_parameters.specificity, 0.88, 1e-7);
+}
```
Comment (reneSchm) [inline cpp/tests/test_abm_testing_strategy.cpp:174]: I don't think adding this test makes sense. The first half only tests whether the default values changed, which is bad unless you have absolute certainty that these defaults are correct and should never change. The second half only checks the basic functionality of a ParameterSet, so this feature is already covered in test_parameter_set.cpp.

**Review by DavidKerkmann — APPROVED** [review]
We already talked about these changes and I am fine with it. We can discuss about the naming "TestData", so if anyone finds a better name, feel free to suggest it.

**Review by reneSchm — CHANGES_REQUESTED** [review]
Good fix, note that the type `mio::CustomIndexArray<mio::abm::TestParameters, mio::abm::TestType>` needs to be bound for the py simulation to run.

File: cpp/models/abm/parameters.h (hunk 1 above)
Comment (khoanguyen-dev) [reply cpp/models/abm/parameters.h:323]: It's now added.

File: cpp/models/abm/test_type.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#ifndef EPI_ABM_TEST_TYPE_H
+#define EPI_ABM_TEST_TYPE_H
```
[the remark this answers is not in the document]
Comment (khoanguyen-dev) [reply cpp/models/abm/test_type.h]: Fixed!

File: cpp/tests/test_abm_testing_strategy.cpp (hunk 2 above)
Comment (khoanguyen-dev) [reply cpp/tests/test_abm_testing_strategy.cpp]: Ok, it was added to verify that we don't have the bug anymore. I've now removed it.

**Comment by khoanguyen-dev** [comment]
There is still an error in the Pycode. The `mio::CustomIndexArray<mio::abm::TestParameters, mio::abm::TestType> ` could not be bound since the `pymio::bind_CustomIndexArray` function does not include an overload for a tuple of doubles (just a single double). 
I will try to work to extend this. If time is essential, I can set up the `TestParameters` in the test directly instead of taking it from the `world.parameters`.

**Comment by reneSchm** [comment]
According to CodeCov the coverage dropped because of some "indirect changes" by commit a8dd034, so now line 200-202 in [cpp/models/abm/infection.cpp](https://app.codecov.io/gh/SciCompMod/memilio/pull/1013/blob/cpp/models/abm/infection.cpp#L200) are no longer covered. Probably some parameter change in the tests could fix this, but maybe have a look at the function as well (it might possibly need updating?).

**Comment by DavidKerkmann** [comment]
It seems that I managed to fix the python bindings. However, I had to implement serialise and deserialise for the simple
```
struct TestParameters {
     UncertainValue<> sensitivity;	     UncertainValue<> sensitivity;
     UncertainValue<> specificity;	     UncertainValue<> specificity;
}
```
as it couldn't be detected by the templates in io/io.h. Is there a way around that? I can imagine letting this derive from std::pair and write access operators for `.first` and `.second` but there may be something more simple. Please comment if you have ideas.

**Comment by reneSchm** [comment]
> Is there a way around that? 

Not really, we cannot iterate over members of generic structs. (Technically, we could using a void\*, but that *will* cause more issues than it solves.)

> I can imagine letting this derive from std::pair and write access operators for .first and .second

That (or tuples, vectors, or other types with its own (de)serialize_internal function) would work, but you lose the names when writing the Json.

I'd say the best options are
- Write the (de)serialize functions. It is not too much work, and does everything you need.
  - I am also working on a shorter serialization feature in the branch connected to branch 652, see [here](https://github.com/SciCompMod/memilio/blob/b1aa14fb64c3f63bbdc14d5f0991a4ef622ea479/cpp/models/abm/parameters.h#L429) for example.
- Replace the TestParameter completely by a std::pair. You can use std::tie or structured binding to set the names, e.g.
```
auto& [sensitivity, specificity] = params.get<GenericTest>();
```

**Comment by DavidKerkmann** [comment]
I see. In that case I would leave it as is for now and make changes later if we need to.

File: cpp/models/abm/parameters.h [hunk 4]
```diff
@@ -314,47 +315,54 @@ struct HighViralLoadProtectionFactor {
...
-        return Type{0.8, 0.88};
-    }
-    static std::string name()
-    {
-        return "AntigenTest";
-    }
+     /**
+     * serialize this. 
+     * @see mio::serialize
+     */
```
Comment (reneSchm) [inline cpp/models/abm/parameters.h:322]:
```suggestion
     /**
      * serialize this. 
      * @see mio::serialize
      */
```

File: cpp/models/abm/parameters.h [hunk 5]
```diff
@@ -314,47 +315,54 @@ struct HighViralLoadProtectionFactor {
...
+     {
+         auto obj = io.create_object("TestParameters");
+         obj.add_element("Sensitivity", sensitivity);
+         obj.add_element("Specificity", specificity);
+     }
+
+     /**
+     * deserialize an object of this class.
+     * @see mio::deserialize
+     */
```
Comment (reneSchm) [inline cpp/models/abm/parameters.h:334]:
```suggestion
     /**
      * deserialize an object of this class.
      * @see mio::deserialize
      */
```

**Review by reneSchm — APPROVED** [review]
Seems like everything works now, good fix for the CustomIndexArray bindings!
Before you merge, could you check the benchmarks? Though I expect there won't be any changes.

**Comment by DavidKerkmann** [comment]
Benchmark comparison:

This branch:
```
Run on (8 X 2300 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 6144 KiB (x1)
Load Average: 3.95, 5.02, 4.52
---------------------------------------------------------------------------
Benchmark                                 Time             CPU   Iterations
---------------------------------------------------------------------------
abm_benchmark/abm_benchmark_50k        2170 ms         1570 ms            1
abm_benchmark/abm_benchmark_100k       4543 ms         3422 ms            1
abm_benchmark/abm_benchmark_200k       9417 ms         7609 ms            1
```

Main:
```
Run on (8 X 2300 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 6144 KiB (x1)
Load Average: 3.15, 3.25, 3.44
---------------------------------------------------------------------------
Benchmark                                 Time             CPU   Iterations
---------------------------------------------------------------------------
abm_benchmark/abm_benchmark_50k        2031 ms         1628 ms            1
abm_benchmark/abm_benchmark_100k       3719 ms         3262 ms            1
abm_benchmark/abm_benchmark_200k       7412 ms         6677 ms            1
```

**Review by reneSchm — APPROVED** [review]

**Pull request merged**